packages/sanity/src/_internal/cli/actions/schema/utils/schemaStoreValidation.ts

```diff
@@ -0,0 +1,235 @@
+import {type CliOutputter} from '@sanity/cli'
+import uniqBy from 'lodash/uniqBy'
+
+import {isDefined} from '../../../../manifest/manifestTypeHelpers'
+import {SANITY_WORKSPACE_SCHEMA_TYPE} from '../../../../manifest/manifestTypes'
+import {type DeleteSchemaFlags} from '../deleteSchemaAction'
+import {type SchemaListFlags} from '../listSchemasAction'
+import {type StoreSchemasFlags} from '../storeSchemasAction'
+import {getManifestDirPath} from './manifestReader'
+import {getStringArrayOutString} from './storeSchemaOutStrings'
+
+export const validForIdChars = 'a-zA-Z0-9._-'
+export const validForIdPattern = new RegExp(`^[${validForIdChars}]+$`, 'g')
+
+const requiredInId = SANITY_WORKSPACE_SCHEMA_TYPE.replace(/[.]/g, '\\.')
+const idPattern = new RegExp(
+  `^(?:[${validForIdChars}]+?\\.)?${requiredInId}\\.([${validForIdChars}]+)$`,
+)
+
+export class FlagValidationError extends Error {
+  constructor(message: string) {
+    super(message)
+    this.name = 'FlagValidationError'
+  }
+}
+
+interface WorkspaceSchemaId {
+  schemaId: string
+  workspace: string
+}
+
+export interface StoreSchemaCommonFlags {
+  'extract-manifest'?: boolean
+  'manifest-dir'?: string
+  'verbose'?: boolean
+}
+
+function parseCommonFlags(
+  flags: StoreSchemaCommonFlags,
+  context: {workDir: string},
+  errors: string[],
+) {
+  const manifestDir = parseManifestDir(flags, errors)
+  const verbose = !!flags.verbose
+  // extract manifest by default: our CLI layer handles both --extract-manifest (true) and --no-extract-manifest (false)
+  const extractManifest = flags['extract-manifest'] ?? true
+
+  const fullManifestDir = getManifestDirPath(context.workDir, manifestDir)
+  return {
+    manifestDir: fullManifestDir,
+    verbose,
+    extractManifest,
+  }
+}
+
+export function parseStoreSchemasConfig(flags: StoreSchemasFlags, context: {workDir: string}) {
+  const errors: string[] = []
+
+  const commonFlags = parseCommonFlags(flags, context, errors)
+  const workspaceName = parseWorkspace(flags, errors)
+  const idPrefix = parseIdPrefix(flags, errors)
+  const schemaRequired = !!flags['schema-required']
+
+  assertNoErrors(errors)
+  return {...commonFlags, workspaceName, idPrefix, schemaRequired}
+}
+
+export function parseListSchemasConfig(flags: SchemaListFlags, context: {workDir: string}) {
+  const errors: string[] = []
+
+  const commonFlags = parseCommonFlags(flags, context, errors)
+  const id = parseId(flags, errors)
+  const json = !!flags.json
+
+  assertNoErrors(errors)
+  return {...commonFlags, json, id}
+}
+
+export function parseDeleteSchemasConfig(flags: DeleteSchemaFlags, context: {workDir: string}) {
```

Comment:
the various parser functions in here seems like good candidates for unit tests, at least the non-trivial ones :)